design/high-level-reactivity.md

```diff
@@ -0,0 +1,245 @@
+
+## Reactive GraphQL
+
+People are excited about GraphQL, and it solves a lot of the problems which many developers, including Meteor’s customers and users, have brought to light. But when you’re building an app, you need more than just a way to query the server and get a response. For example, some parts of your app will want the data to update reactively when a different user makes a change.
+
+The client and server parts of the GraphQL system will need to collaborate to make this possible, but one of our goals is to make minimal changes to the current reference implementation of GraphQL so that developers can take advantage of current and future productivity tools for GraphQL.
+
+### Dependencies
+
+The core concept of the system we envision is the idea of a dependency, or dep for short. A dep is a tuple of `key` and `version`, where `key` globally represents a particular unit of data (for example, a row in a database or the result of a query), and `version` globally represents how many times it has been changed.
+
+The intended result is that if you are holding a piece of data, and you have a dep for it, you can ask the global dependency server whether it has changed by sending the key and version. The main benefit of this approach is that the application server doesn't need to be aware of which queries each client is currently tracking - the client can keep that knowledge for itself, reducing the burden on the server and giving the developer more options for controlling when new data is fetched.
+
+### Returning dependencies to the client
+
+Before, we gave an example of a GraphQL response for a simple query. If you make a reactive GraphQL query, you will get extra metadata about the dependencies involved in particular subtrees, like so:
+
+```
+{
+  me: {
+    username: "sashko",
+    lists: [
+      {
+        id: 1,
+        name: "My first todo list",
+        tasks: [ ... and so on ],
+        _deps: {
+          _self: { key: '12341234', version: 3 },
+          tasks: { key: '35232345', version: 4 }
+        }
+      }
+    ]
+  }
+}
+```
+
+This will tell the client which dependencies they should watch to figure out if the list object itself changed, or if the list of tasks needs to be updated.
+
+Note that can’t put the `_deps` field on tasks itself, because JSON doesn’t allow it, so we have to put it on the parent instead. Also, the `_self` field is a shorthand to avoid listing all of the properties of the list object - it could have `name`, `description`, and so on, and it would be a waste of bandwidth to send all of those key names again.
+
```

Comment:
I think it would be something in the data container. There's a place in Relay where it adds the `id` field even if you didn't request it, so we should just hop on that. The Relay team helpfully pointed out that location, so I'll try to do something there. It's going to likely be a public API on the Relay store in the future.